### PR TITLE
Edit git-tar to deploy functions in parallel

### LINKED ISSUE #252: git-tar should invoke deploy (buildshiprun) in parallel
## Expected Behaviour

git-tar should invoke deploy (buildshiprun) in parallel

## Current Behaviour

If a stack has 10 functions they are built in serial one at a time. Use an errgroup (see of-builder for an example of this) to build them parallel, but with a limit applied so that the concurrency is manageable - i.e. 2 or 4 at once.

The errgroup can be used to synchronise and collect any errors that come up.

Serial deployment loop:

https://github.com/openfaas/openfaas-cloud/blob/master/git-tar/function/ops.go#L319

errgroup usage:

https://github.com/openfaas/openfaas-cloud/blob/master/of-builder/main.go#L196

### PR DESCRIPTION
**Pull request by ivanayov**
With this change git-tar will POST to buildshiprun in parallel,
with a restriction on maximum 4 functions to be processed at once
(If required, this can be made configurable)

It will wait for all functions to be deployed, but will return the first error only (this is how errgroup behaves). The reason to implement it this way is that it will behave as a typical build system. Second if someone has multiple functions in a single repo, this means that they should be dependent on each other and reporting the first error and fixing the functions one by one in this context makes sense.
The other option is to keep back `failedFunctions` and synchronise it, but it's aways better not to add synchronised array when it's possible, given this opens a field for future bugs.

Signed-off-by: Ivana Yovcheva <iyovcheva@vmware.com>

Closes #252 , #93

## How Has This Been Tested?

Tested on this repo - [ivanayov/ofc-pr-testing/runs/26567991](https://github.com/ivanayov/ofc-pr-testing/runs/26567991)
The functions were building in parallel and the status was updated successfully at the end.
With one function failing, it waits for all and reports failing status for that one.
With more than one failing - the same, but reports one of them as failing.

For repo with 1 function, works as normal, but starts the function build few seconds later.

For repo with 0 functions fails with `unable to find stack.yml`

## How are existing users impacted? What migration steps/scripts do we need?

Not affected

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [x] read the [CONTRIBUTION](https://github.com/openfaas/faas/blob/master/CONTRIBUTING.md) guide
- [x] signed-off my commits with `git commit -s`
- [ ] added unit tests - not compatible for the use case, needs live testing instead

### CODE REVIEW

File: git-tar/function/ops.go [hunk 1]
```diff
@@ -316,91 +317,98 @@ func deploy(tars []tarEntry, pushEvent sdk.PushEvent, stack *stack.Services, sta
 	c := http.Client{}
 	gatewayURL := os.Getenv("gateway_url")
 
+	sem := make(chan int, 4)
```
Comment (alexellis) [inline git-tar/function/ops.go]: sem?

File: git-tar/function/ops.go [hunk 2]
```diff
@@ -316,91 +317,98 @@ func deploy(tars []tarEntry, pushEvent sdk.PushEvent, stack *stack.Services, sta
...
+	eg, _ := errgroup.WithContext(context.Background())
+
 	for _, tarEntry := range tars {
-		fmt.Println("Deploying service - " + tarEntry.functionName)
+		sem <- 1
+		eg.Go(func() error {
```
Comment (alexellis) [inline git-tar/function/ops.go]: This is too nested now, I'd like to see it broken down and broken out.

File: git-tar/function/ops.go (hunk 1 above)
Comment (ivanayov) [reply git-tar/function/ops.go]: This is for semaphore. Does it look clear or I should better rename?

Comment (alexellis) [reply git-tar/function/ops.go]: It's not a semaphore it's a channel. Consider a more meaningful name. workChannel? (I don't know)

File: git-tar/function/ops.go (hunk 2 above)
Comment (ivanayov) [reply git-tar/function/ops.go]: It's addressed on next lines

File: git-tar/function/ops.go [hunk 3]
```diff
@@ -306,103 +308,122 @@ func clone(pushEvent sdk.PushEvent) (string, error) {
...
-	failedFunctions := []string{}
-	owner := pushEvent.Repository.Owner.Login
-	repoName := pushEvent.Repository.Name
-	url := pushEvent.Repository.CloneURL
-	afterCommitID := pushEvent.AfterCommitID
-	installationID := pushEvent.Installation.ID
-
-	c := http.Client{}
-	gatewayURL := os.Getenv("gateway_url")
+	sem := make(chan int, 4)
```
Comment (alexellis) [inline git-tar/function/ops.go]: Let's get this variable name changed please @ivanayov

**Comment by ivanayov** [comment]
@alexellis I've addressed your comment.
The PR now is fixed

**Comment by alexellis** [comment]
Needs rebasing and another review.

File: git-tar/function/ops.go [hunk 4]
```diff
@@ -306,109 +308,128 @@ func clone(pushEvent sdk.PushEvent) (string, error) {
...
-	url := pushEvent.Repository.CloneURL
-	afterCommitID := pushEvent.AfterCommitID
-	installationID := pushEvent.Installation.ID
-	sourceManagement := pushEvent.SCM
-	privateRepo := pushEvent.Repository.Private
-	repositoryURL := pushEvent.Repository.RepositoryURL
-
-	c := http.Client{}
-	gatewayURL := os.Getenv("gateway_url")
+	workCh := make(chan int, 4)
```
Comment (alexellis) [inline git-tar/function/ops.go]: `4` should be a const and also let's set it to `2` so we don't cripple the system.

**Comment by ivanayov** [comment]
@rosti thanks for the code review. Can you please add a comment?

**Comment by rosti** [comment]
The current version looks OK from concurrency point of view.

The only obvious things are a bit stylistic. I see that you use the `x, _ := ...` construct very often. You can reduce that to just `x := ...`.

**Comment by ivanayov** [comment]
@rosti this syntax does not work for function call

**Comment by ivanayov** [comment]
There's an issue with GitHub I think. Travis build is [passing](https://travis-ci.org/openfaas/openfaas-cloud/builds/451913906#L1654), but the status stays as pending for ~24h now.

**Comment by burtonr** [comment]
@ivanayov I re-ran the build in Travis, and the check was updated successfully!

Just need to ping @alexellis to make sure his changes were addressed so this can be merged

**Comment by alexellis** [comment]
Bumping up the issue as @LucasRoesler mentioned this recently. 

@ivanayov please do an e2e test in a live cluster using repos with 0, 1 and multiple functions.

**Comment by alexellis** [comment]
Derek add label: blocked

**Comment by alexellis** [comment]
Blocked awaiting more testing.

**Comment by ivanayov** [comment]
I'm on it

**Comment by alexellis** [comment]
Given the potential issues found, let's park this for the time being.

**Comment by alexellis** [comment]
Derek close: due to issues found during testing.

**Pull request closed without merging**